packages/blocks/src/api/utils.js

```diff
@@ -121,3 +123,87 @@ export function normalizeBlockType( blockTypeOrName ) {
 
 	return blockTypeOrName;
 }
+
+/**
+ * Get the label for the block, usually this is either the block title,
+ * or the value of the block's `label` function when that's specified.
+ *
+ * @param {Object} blockType  The block type.
+ * @param {Object} attributes The values of the block's attributes.
+ * @param {Object} context    The intended use for the label.
+ *
+ * @return {string} The block label.
+ */
+export function getBlockLabel( blockType, attributes, context = 'visual' ) {
+	const {
+		__experimentalLabel: getLabel,
+		title,
+	} = blockType;
+
+	const label = getLabel && getLabel( attributes, { context } );
+
+	if ( ! label ) {
+		return title;
+	}
+
+	// Strip any HTML (i.e. RichText formatting) before returning.
+	return stripHTML( label );
+}
+
+/**
+ * Get a label for the block for use by screenreaders, this is more descriptive
+ * than the visual label and includes the block title and the value of the
+ * `getLabel` function if it's specified.
+ *
+ * @param {Object}  blockType  The block type.
+ * @param {Object}  attributes The values of the block's attributes.
+ * @param {?number} row        The row of the block in the block list.
+ *
+ * @return {string} The block label.
+ */
+export function getAccessibleBlockLabel( blockType, attributes, row ) {
+	// `title` is already localized, `label` is a user-supplied value.
+	const { title } = blockType;
+	const label = getBlockLabel( blockType, attributes, 'accessibility' );
+	const hasRow = row !== undefined;
+
+	// getBlockLabel returns the block title as a fallback when there's no label,
+	// if it did return the title, this function needs to avoid adding the
+	// title twice within the accessible label. Use this `hasLabel` boolean to
+	// handle that.
+	const hasLabel = label && label !== title;
+
+	if ( hasRow ) {
+		if ( hasLabel ) {
+			return sprintf(
+				/* translators: accessibility text. %s: block title, %d block row number, $s block label.. */
+				__( '%s Block. Row %d. %s' ),
```

Comment:
TODO, add position indicators to the localisation string - '%1$s Block. Row %2$d. %3$s' and update the comments.